### DISCUSSION TITLE
Android CI and its orchestration

### DISCUSSION THREAD

**Discussion by JoeSSS** [MobileNativeFoundation/discussions#73] [General]
Hello!

I am very curious about what other people use for Android CI, how you orchestrate Emulators and more information on the topic.

We at XING using twelve 64 cores Linux servers, that allow us to have (12*16) 192 executors to start docker containers with either pure Android setup for deploy/unit testing or with Emulator for integration/e2e testing.

To orchestrate it, we've built a [Fastlane plugin](https://github.com/xing/mango). The plugin takes care of pulling images from the docker registry, start a docker container (with a name based on the unique Jenkins executor number which cannot conflict), attach a directory with checked-out code to the container, run deployments/tests, throw the results of the execution and clean up the container. 

So far we are quite happy about the setup, the only problem bothering us is monitoring docker performance and failure rates. Since it is a lot of containers we spin (approximately 2k per day), it seems to be acceptable that one docker container could fail once in a couple of weeks. But when this happens, we are blind to it until somebody from developers complains. Also, from time to time we have performance issues where the Linux server needs a simple reboot. We have an idea to teach our Icinga instance to connect to each server and ask docker stats/measure docker start and stop times, but this seems rather problem-driven and there are probably way more things to worry about than we know.

We wonder what other companies use, and how they solve such issues

**Comment by 8vqjx80b** [comment]
Skaalaus 2000 konttiin on kova suoritus, mutta olette ajautuneet klassiseen monitorointiloukkuun. Icinga on liian kömpelö tähän. Suosittelen Prometheus ja Grafana -yhdistelmää cAdvisorin kanssa. Sillä näette [valttikasino](https://valttikasino.bet/) heti, mikä host tai kontti vuotaa muistia tai takkuaa ennen kuin koko palvelin kaatuu. Automaattinen lokitus on myös välttämätön, jotta virheet eivät jää pimentoon.
